OPENGL: Support for shader-based scalers

sev- opened the pull request:
Hi team,

Finally, I can provide the RFC for the shader-based scaler support.

The work has been stared by @lordhoto back in 2016: https://github.com/lordhoto/scummvm/tree/libretro-shader

Then, as you see from the history, was picked up by @lotharsm, and finally, @ccawley2011 implemented the initial version of GUI. Then I just wrapped up their work and implemented the fallback, test and ensured it works well in different configurations.

The scalers which are meant to be used are GLSL scalers from RetroArch project: https://github.com/libretro/glsl-shaders

What we are yet to do is select those shaders which do work and are meaningful and ship them together with ScummVM.

When Shaders are selected, then I force the backend to Normal1x scaler, so it does not go in the way. Also, I had to disable Filtering because it is shader-based as well and goes in direct conflict with the shader. Moreover, it makes little sense with this new setup.

The new GUI:

Selecting Scalers:
<img width="915" alt="Screenshot 2022-10-05 at 16 45 04" src="https://user-images.githubusercontent.com/365188/194091233-d4227417-c441-4193-8514-3060ac459e64.png">

Selecting Shaders:
<img width="915" alt="Screenshot 2022-10-05 at 16 44 53" src="https://user-images.githubusercontent.com/365188/194091238-06568a61-1804-4a79-826f-b24c71ffc8fe.png">

Once a new shader is selected, the user is presented with a test pattern. This is achieved by initialising the game surface and drawing the pattern on it, this way simulating the exact in-game experience. The chosen pattern is Philips PM5544, a standard used in old television. I chose it because it also clearly shows geometric distortions, if any.

Also, the scalers may not work. If they cannot compile, the user is presented with an error message, or sometimes, the scaler may not yield visible results. This is why a Windows-like countdown is shown and the settings are reverted if the user does not confirm.

<img width="915" alt="Screenshot 2022-10-05 at 16 45 36" src="https://user-images.githubusercontent.com/365188/194091225-b88b47f6-e544-406a-b164-22ea550de3c8.png">

And in-game it looks like this:
<img width="915" alt="Screenshot 2022-10-05 at 16 46 03" src="https://user-images.githubusercontent.com/365188/194091186-3ab4de6c-1c09-4270-85a5-0a122e5da085.png">

We were brainstorming the UI with @Thunderforge, but there are a bunch of technical limitations:
1. It is not possible to apply scaler to a given surface, so generating a preview is not possible
2. We were thinking about bundling shaders in the executable and presenting them in a popup. This could require two separate ways of specifying the shaders as internal and external ones and is not portable.

Thus, the current GUI is a technical compromise.

Please, test. A couple of scalers I tested with: crt/fake-CRT-Geom.glslp, crt/crt-hyllian-fast.glslp

Many other scalers do not work. I could take a look into some of the compilation errors there, but this is the primary reason why we need to do the work on pre-selecting them.

ccawley2011 commented:
Looking good so far. A few initial observations:

> When Shaders are selected, then I force the backend to Normal1x scaler, so it does not go in the way.

There was some discussion in the past about combining shaders and scalers, so it might be worth allowing this: https://forums.scummvm.org/viewtopic.php?p=88798#p88798

> Also, I had to disable Filtering because it is shader-based as well and goes in direct conflict with the shader. Moreover, it makes little sense with this new setup.

It would still be useful for OpenGL 1.x and OpenGL ES 1.1 setups where shader support is unavailable. It should be possible to disable filtering for individual textures and/or targets specifically when shaders are in use, regardless of the setting. I can look into this if needed.

> This is achieved by initialising the game surface and drawing the pattern on it, this way simulating the exact in-game experience.

Would it be worth allowing a test card graphic to be specified in the theme files as well?

sev- commented:
> There was some discussion in the past about combining shaders and scalers, so it might be worth allowing this: https://forums.scummvm.org/viewtopic.php?p=88798#p88798

Hmm, I was not anticipating this and spent a sizeable effort on implementing their mutual exclusivity.... I can revert it.

> > Also, I had to disable Filtering because it is shader-based as well and goes in direct conflict with the shader. Moreover, it makes little sense with this new setup.
> 
> It would still be useful for OpenGL 1.x and OpenGL ES 1.1 setups where shader support is unavailable. It should be possible to disable filtering for individual textures and/or targets specifically when shaders are in use, regardless of the setting. I can look into this if needed.

In this case, it would make sense to add couple of `#ifdef`  and report that the OpenGL 1.x support filtering but not shaders. I need help with these defines. Any hints how I can determine it run-time or compile-time?

> > This is achieved by initialising the game surface and drawing the pattern on it, this way simulating the exact in-game experience.
> 
> Would it be worth allowing a test card graphic to be specified in the theme files as well?
Currently, I generate the pattern programmatically. See `drawTestPattern()` in `gui/options.cpp`

sev- commented:
Okay, I reverted the change with the radio buttons. Now it is possible again to use both scalers and shaders at the same time. The update GUI:
<img width="915" alt="Screenshot 2022-10-06 at 15 16 37" src="https://user-images.githubusercontent.com/365188/194323031-cf2942e4-a02a-4cb2-b790-5ea2ac3ce1f0.png">

sev- commented:
@ccawley2011 Do I understand correctly that those OpenGL platforms could be caught by `#if USE_FORCED_GLES`?

lephilousophe commented:
> @ccawley2011 Do I understand correctly that those OpenGL platforms could be caught by `#if USE_FORCED_GLES`?
That's the static detection part.
You still have to check dynamically with:
```
if (g_system->getOpenGLType() == OpenGL::kContextGLES) {
```

sev- commented:
@lephilousophe Thanks!

I now implemented showing Filtering checkbox on those platforms, fixed a crash when shaders are not supported and properly hide the Filtering checkbox when it is not present.

At this moment, I have all features implemented. What I possibly will take a look at is those vertical artifacts on the test pattern.

Looking forward to your next round of reviews.

lephilousophe commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -100,13 +109,19 @@ bool OpenGLGraphicsManager::hasFeature(OSystem::Feature f) const {
+#if !USE_FORCED_GLES
+	case OSystem::kFeatureShaders:
 #endif
 		return true;
```
You should do: `return LibRetroPipeline::isSupportedByContext()` here

lephilousophe commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -100,13 +109,19 @@ bool OpenGLGraphicsManager::hasFeature(OSystem::Feature f) const {
 
+	case OSystem::kFeatureFilteringMode:
+		// Filtering is supported only in cases when shaders are not present
+		return (g_system->getOpenGLType() == OpenGL::kContextGLES);
```
Here: `return !LibRetroPipeline::isSupportedByContext()`

lephilousophe commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -123,6 +138,10 @@ void OpenGLGraphicsManager::setFeatureState(OSystem::Feature f, bool enable) {
 
 	case OSystem::kFeatureFilteringMode:
+		// Filtering is supported only in cases when shaders are not present
+		if (g_system->getOpenGLType() != OpenGL::kContextGLES)
```
Here too: `return !LibRetroPipeline::isSupportedByContext()`

lephilousophe reviewed (commented):
Sorry, I just did a dumb reply because I read about GLES.
After looking at the code, I have a better suggestion.

sev- commented:
There was on overlook: the attribution to the first commit by @lordhoto got missed in translation. I fixed this and brining my apologies to Johannes, this feature wouldn't be possible without his initial work.

lordhoto commented:
> There was on overlook: the attribution to the first commit by @lordhoto got missed in translation. I fixed this and brining my apologies to Johannes, this feature wouldn't be possible without his initial work.

Thanks you Eugene, Glad some people finally picked this up and made it work. This is hopefully a really cool feature for ScummVM users.

athrxx commented on backends/graphics/opengl/pipelines/libretro.cpp:
```diff
@@ -0,0 +1,541 @@
+
+		Common::SeekableReadStream *stream = fileNode.createReadStream();
+		if (!stream) {
+			warning("LibRetroPipeline::loadPasses: Could not open file '%s'", fileNode.getName().c_str());
```
It will trigger an assert if the filenode is invalid. Maybe replace the `'fileNode.getName().c_str()` with `fileNode.isReadable() ? fileNode.getName().c_str() : i->fileName.c_str()`...

(test case: cel/presets/MMJ_Cel_Shader_3dfx.glslp)

sev- replied:
Good point, done

elasota commented:
Seems like FilteringMode isn't used anywhere?

ccawley2011 commented on backends/graphics/opengl/pipelines/libretro/parser.cpp:
```diff
@@ -0,0 +1,491 @@
+			continue;
+
+		// Split line into key, value pair.
+		// TODO: Files can contain comments starting with '#', we need to
+		// handle this.
```
Should this be removed now that we support comments?

sev- replied:
Done

ccawley2011 commented on gui/options.cpp:
```diff
@@ -3564,4 +3606,306 @@ void GlobalOptionsDialog::storageErrorCallback(Networking::ErrorResponse respons
+	circleColorNoblend(surface, c1x, c1y, gapsize + 1, WHITE, false);
+}
+
+static void drawTestPattern() {
```
Would it be possible to have this in a separate file? It looks like it could be useful in the Testbed engine as well.

sev- replied:
Indeed. Moved to `graphics/pm5544.h`

sev- commented:
> Seems like FilteringMode is defined but isn't actually used anywhere?

Initially, that was used for OpenGL1 and OpenGL ES 1.0/1.1. But now, after merging @ccawley2011 PR, it is back to its former glory.

bluegr commented:
Great work, very nice feature! Kudos to everyone involved in designing and implementing it.
I like the way this has been implemented, and it's good that this feature isn't mutually exclusive with the scaling.
+1 from me

sev- commented:
Thanks, merging.
